### PR TITLE
[KYUUBI #544] Add UT for engine deregister exception and refactor package name for SparkSQLEngineListenerSuite

### PR DESCRIPTION
**Pull request by turboFei**
### _Why are the changes needed?_

Add UT for engine deregister exception

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.readthedocs.io/en/latest/tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/spark/kyuubi/SparkSQLEngineDeregisterSuite.scala [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    assert(engine.discoveryService.getServiceState === ServiceState.STARTED)
+    intercept[SparkException](spark.sql(query).collect())
+    assert(engine.discoveryService.getServiceState === ServiceState.STOPPED)
+  }
+}
+
+class SparkSQLEngineDeregisterExceptionSuite extends SparkSQLEngineDeregisterSuite {
+  override def withKyuubiConf: Map[String, String] = {
+    super.withKyuubiConf ++ Map(ENGINE_DEREGISTER_EXCEPTION_CLASSES.key ->
+      classOf[SparkException].getCanonicalName)
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/spark/kyuubi/SparkSQLEngineDeregisterSuite.scala:52]: can we improve this to check the root cause？

File: externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/spark/kyuubi/SparkSQLEngineDeregisterSuite.scala [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import java.util.UUID
+
+import org.apache.spark.SparkException
+import org.apache.spark.sql.internal.SQLConf.{ANSI_ENABLED, DECIMAL_OPERATIONS_ALLOW_PREC_LOSS}
+
+import org.apache.kyuubi.config.KyuubiConf._
+import org.apache.kyuubi.engine.spark.WithDiscoverySparkSQLEngine
+import org.apache.kyuubi.service.ServiceState
+
+abstract class SparkSQLEngineDeregisterSuite extends WithDiscoverySparkSQLEngine{
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/spark/kyuubi/SparkSQLEngineDeregisterSuite.scala:29]: space

File: externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/spark/kyuubi/SparkSQLEngineDeregisterSuite.scala (hunk 1 above)
Comment (turboFei) [reply externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/spark/kyuubi/SparkSQLEngineDeregisterSuite.scala:55]: yeah, I plan to improve it.

**Comment by turboFei** [comment]
will simplify the UT.

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SparkSQLEngineListener.scala [hunk 3]
```diff
@@ -18,6 +18,7 @@
 // Some object likes `JobFailed` is only accessible in org.apache.spark package
 package org.apache.spark.kyuubi
 
+import org.apache.spark.SparkException
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SparkSQLEngineListener.scala:21]: The PR changes touched code out scope of `test`, so remove the `[TEST]` in title and rename title to match the real changes.

**Review by pan3793 — APPROVED** [review]
LGTM

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SparkSQLEngineListener.scala [hunk 4]
```diff
@@ -45,21 +46,29 @@ class SparkSQLEngineListener(server: Serverable) extends SparkListener with Logg
   override def onJobEnd(jobEnd: SparkListenerJobEnd): Unit = {
    jobEnd.jobResult match {
      case JobFailed(e) if e != null =>
+       // get valuable exception
+       val ve = e match {
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SparkSQLEngineListener.scala:50]: Please refer to KyuubiSQLException.findCause, we can make one here

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SparkSQLEngineListener.scala [hunk 5]
```diff
@@ -45,21 +46,29 @@ class SparkSQLEngineListener(server: Serverable) extends SparkListener with Logg
...
+         case se: SparkException if se.getCause != null =>
+           se.getCause
+         case _ => e
+       }
+
        var deregisterInfo: Option[String] = None
-       if (deregisterExceptions.exists(_.equals(e.getClass.getCanonicalName))) {
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SparkSQLEngineListener.scala:49]: On second thought, we can merge message and stack trace together with `KyuubiSQLException.stringlifyException` and check once? we can save one config for better user experience.

WDYT @pan3793

Comment (pan3793) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SparkSQLEngineListener.scala:49]: I'm ok with it, but how to name this conf?

Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SparkSQLEngineListener.scala:49]: we can keep the `message` as is, update the config doc and drop another

Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SparkSQLEngineListener.scala:49]: Can you make a separate PR for this? @turboFei

Comment (turboFei) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SparkSQLEngineListener.scala:49]: sure

**Pull request closed without merging**